### PR TITLE
DBAAS-1107: enhance DB instance provisioning for RDS

### PR DESCRIPTION
**Pull request by xieshenzh**
DB instance provisioning changes for RDS as follow up to https://github.com/RHEcosystemAppEng/dbaas-operator/pull/319

Jira: https://issues.redhat.com/browse/DBAAS-1107

### CODE REVIEW

File: bundle.Dockerfile [hunk 1]
```diff
@@ -6,7 +6,7 @@ LABEL operators.operatorframework.io.bundle.manifests.v1=manifests/
 LABEL operators.operatorframework.io.bundle.metadata.v1=metadata/
 LABEL operators.operatorframework.io.bundle.package.v1=rds-dbaas-operator
 LABEL operators.operatorframework.io.bundle.channels.v1=alpha
-LABEL operators.operatorframework.io.metrics.builder=operator-sdk-v1.19.0+git
```
Comment (tchughesiv) [inline bundle.Dockerfile:9]: seems you used newer sdk? should we account for this in Makefile? or at least in README?

Comment (xieshenzh) [reply bundle.Dockerfile:9]: Updated Makefile in this commit: https://github.com/RHEcosystemAppEng/rds-dbaas-operator/pull/16/commits/d64970a430f0a6b97f2c205fd9251f844d8a54e6

I will cherry-pick this commit after updating the dbaas-operator dependency and before make the PR ready for review.

Comment (tchughesiv) [reply bundle.Dockerfile:9]: ah, ok... nice, thanks

File: controllers/db_utils.go [hunk 2]
```diff
@@ -101,6 +251,39 @@ func getDefaultDBName(engine string) *string {
...
+	case postgres:
+		return pointer.String("13.7")
+	case sqlserverEe, sqlserverSe, sqlserverEx, sqlserverWeb:
+		return pointer.String("15.00.4236.7.v1")
+	case oracleSe2, oracleSe2Cdb:
+		return pointer.String("19.0.0.0.ru-2022-10.rur-2022-10.r1")
+	case mysql:
+		return pointer.String("8.0.28")
+	case mariadb:
+		return pointer.String("10.6.10")
```
Comment (jeremyary) [inline controllers/db_utils.go]: how often do you feel these are going to be subject to change? Often enough to be worth of pulling out to vars?

File: controllers/dbaasprovider_controller.go [hunk 3]
```diff
@@ -110,200 +111,3277 @@ func (r *DBaaSProviderReconciler) Reconcile(ctx context.Context, req ctrl.Reques
...
+		return ctrl.Result{}, err
+	}
+
+	if len(clusterRoleList.Items) < 1 {
+		err := errors.NewNotFound(
+			schema.GroupResource{Group: "rbac.authorization.k8s.io", Resource: "ClusterRole"}, "potentialOwner")
+		logger.Error(err, "could not find ClusterRole owned by CSV to inherit operand")
+		return ctrl.Result{}, err
+	}
+
```
Comment (jeremyary) [inline controllers/dbaasprovider_controller.go:126]: Not saying it's wrong, just checking - It's ok to lose the 'if' check wrapping this? We don't mind always performing this bit when we pass through here?

File: controllers/dbaasprovider_controller.go [hunk 4]
```diff
@@ -110,200 +111,3277 @@ func (r *DBaaSProviderReconciler) Reconcile(ctx context.Context, req ctrl.Reques
...
+						DefaultValue: postgres,
+					},
 				},
-				{
-					Name:        dbSubnetGroupName,
-					DisplayName: "Subnet group",
-					Type:        "string",
-					Required:    false,
+			},
+			dbaasoperator.ProvisioningMachineType: {
```
Comment (jeremyary) [inline controllers/dbaasprovider_controller.go]: if we end up doing a ton of maintenance on these lists, we could consider keeping a yaml file to embed or something else to try & ease the pain? Not sure, hopefully we won't need to touch often.

File: controllers/dbaasprovider_controller.go (hunk 3 above)
Comment (xieshenzh) [reply controllers/dbaasprovider_controller.go:126]: Do you mean `if len(clusterRoleList.Items) < 1`?
The code needs one `clusterrole` to [generate](https://github.com/RHEcosystemAppEng/rds-dbaas-operator/blob/7d573c4ec5e0ac32fa7d9f1861c52a1ef0f72f6f/controllers/dbaasprovider_controller.go#L159-L160) the dbaasprovider CR.
This solution is not designed by me orginally, other provider [operators](https://github.com/CrunchyData/crunchy-bridge-operator/blob/e55bad94135a1293e2316c7eb1656b2ac31d4b08/controllers/dbaas.redhat.com/dbaasprovider_reconciler.go#L174-L175) works like this as well.

File: controllers/dbaasprovider_controller.go (hunk 4 above)
Comment (xieshenzh) [reply controllers/dbaasprovider_controller.go]: Moved to yaml file.

File: controllers/db_utils.go (hunk 2 above)
Comment (xieshenzh) [reply controllers/db_utils.go]: Moved the strings to constant variables.
AWS started to support these db versions in 2022. We can use them as long as AWS supports them. We don't have to upgrade to the latest versions unless there are specific requirements.

File: controllers/dbaasprovider_controller.go (hunk 3 above)
Comment (jeremyary) [reply controllers/dbaasprovider_controller.go:126]: I meant where this full block was previously wrapped by a Get check and only followed these steps when NotFound, however, I pulled the full file to compare and see where the predicate means we should encounter that scenario in the controller.

File: controllers/dbaasprovider_controller.go [hunk 5]
```diff
@@ -110,200 +111,3277 @@ func (r *DBaaSProviderReconciler) Reconcile(ctx context.Context, req ctrl.Reques
...
+
 	instance := &dbaasoperator.DBaaSProvider{
 		ObjectMeta: metav1.ObjectMeta{
 			Name: providerCRName,
 		},
 	}
-	if err := r.Get(ctx, client.ObjectKeyFromObject(instance), instance); err != nil {
-		if errors.IsNotFound(err) {
-			// CR deleted since request queued, child objects getting GC'd, no requeue
-			logger.Info("resource not found, creating now")
```
Comment (jeremyary) [inline controllers/dbaasprovider_controller.go:121]: this Get check just  to clarify

File: controllers/dbaasprovider_controller.go (hunk 4 above)
Comment (jeremyary) [reply controllers/dbaasprovider_controller.go]: There's an alternative method for accomplishing this that we've used in a few places in the operator taking advantage of 'embed' to load the YAML file into a []byte variable that'll unmarshal and not have to worry about the reader. We've used it in a few other places so far. 

Check this out  - https://pkg.go.dev/embed

setup:
https://github.com/RHEcosystemAppEng/dbaas-operator/blob/main/controllers/reconcilers/quickstartinstallation/quickstart_installation_reconciler.go#L21-L31

use:
https://github.com/RHEcosystemAppEng/dbaas-operator/blob/main/controllers/reconcilers/quickstartinstallation/quickstart_installation_reconciler.go#L68-L69

~~WDYT about swapping to this for consistency? Open to hearing reasons if not a fan, it's still a new tool. :)~~

edit: ehh, thinking on it, it's not like we're in the same codebase, so maybe just a new toy to check out and consider

**Review by jeremyary — APPROVED** [review]

File: controllers/dbaasprovider_controller.go (hunk 3 above)
Comment (xieshenzh) [reply controllers/dbaasprovider_controller.go:126]: The code is changed to do a `CreateOrUpdate`. So if the dbaasprovider object exists, the code will update it, which is necessary when we need to change the dbaasprovider object for upgrade.

File: controllers/dbaasprovider_controller.go (hunk 4 above)
Comment (xieshenzh) [reply controllers/dbaasprovider_controller.go]: Will change it to take this approach later.

**Pull request merged**